### PR TITLE
[Do Not Merge][Client] Manual keep alive pings

### PR DESCRIPTION
**Pull request by ckw017**
## Why are these changes needed?

Manual keep alive pings for data/log client every 10 seconds. KeepAlive's are forwarded through proxy to specific instances

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [x] This PR is not tested :(

### CODE REVIEW

File: python/ray/util/client/server/dataservicer.py [hunk 1]
```diff
@@ -143,6 +143,11 @@ def Datapath(self, request_iterator, context):
                         logger.debug("Shutting down ray.")
                         ray.shutdown()
 
+    def KeepAlive(self, request, context):
```
Comment (ijrsvt) [inline python/ray/util/client/server/dataservicer.py:146]: Can you add some types here?

File: python/ray/util/client/dataclient.py [hunk 2]
```diff
@@ -59,9 +68,12 @@ def _next_id(self) -> int:
     def _start_datathread(self) -> threading.Thread:
         return threading.Thread(target=self._data_main, args=(), daemon=True)
 
+    def _start_keepalive_thread(self) -> threading.Thread:
+        return threading.Thread(
+            target=self._keepalive_main, args=(), daemon=True)
```
Comment (ijrsvt) [inline python/ray/util/client/dataclient.py:73]: Hmm I wonder if we can have some common `_run_keepalive` function that takes in an event & a stub? That way the dataclient & logclient can share this function!

File: python/ray/util/client/server/proxier.py [hunk 3]
```diff
@@ -473,30 +473,40 @@ def Datapath(self, request_iterator, context):
...
         self.proxy_manager = proxy_manager
 
-    def Logstream(self, request_iterator, context):
-        client_id = _get_client_id_from_context(context)
-        if client_id == "":
-            return
-        logger.debug(f"New logstream connection from client {client_id}: ")
-
-        channel = None
+    def _try_get_channel(self, client_id):
```
Comment (ijrsvt) [inline python/ray/util/client/server/proxier.py:490]: nit: typing of `client_id`

**Review by ijrsvt — COMMENTED** [review]
Quick question, does this handle Proxier <-> Specific Server timeouts?

File: python/ray/util/client/dataclient.py [hunk 4]
```diff
@@ -100,11 +112,43 @@ def _data_main(self) -> None:
                 logger.exception(
                     "Got Error from data channel -- shutting down:")
 
+    def _keepalive_main(self) -> None:
```
Comment (AmeerHajAli) [inline python/ray/util/client/dataclient.py:115]: can we add a function level description please?

File: python/ray/util/client/logsclient.py [hunk 5]
```diff
@@ -60,6 +72,35 @@ def _log_main(self) -> None:
                 logger.exception(
                     f"Got Error from logger channel -- shutting down: {e}")
 
+    def _keepalive_main(self) -> None:
```
Comment (AmeerHajAli) [inline python/ray/util/client/logsclient.py:75]: This looks similar to the previous main with some differences in warning/wait time? can we unify the two functions?

File: python/ray/util/client/dataclient.py [hunk 6]
```diff
@@ -19,6 +21,9 @@
 
 ResponseCallable = Callable[[ray_client_pb2.DataResponse], None]
 
+# How often to make manual KeepAlive calls
+DATACLIENT_KEEPALIVE_INTERVAL = 30
```
Comment (AmeerHajAli) [inline python/ray/util/client/dataclient.py:25]: should we make it less than the product's interval which IIRC is set to 15?
@ijrsvt knows more, most likely I am wrong here.

**Comment by ckw017** [comment]
> Quick question, does this handle Proxier <-> Specific Server timeouts?

@ijrsvt I think it should, it the proxy forwards the keepalive request to the actual logclient/dataclient. Logic for that is [here](https://github.com/ckw017/ray/blob/manualkeepalive/python/ray/util/client/server/proxier.py#L524-L530) and [here](https://github.com/ckw017/ray/blob/manualkeepalive/python/ray/util/client/server/proxier.py#L476-L482)

File: python/ray/util/client/dataclient.py (hunk 6 above)
Comment (ckw017) [reply python/ray/util/client/dataclient.py]: Set to 10 seconds for now just to be safe.

**Comment by ckw017** [comment]
> Thanks @ckw017 , this is looking really good. Is it possible to unit test this somehow? Can you try to test it manually?

Tested manually for now (checked logs for specific log/data servicer to make sure keepalives were reaching them/being returned properly to the clients). Can look into a unit test in a follow-up PR if this does turn out to be a solution for the disconnect isssue

File: python/ray/util/client/server/proxier.py [hunk 7]
```diff
@@ -487,7 +488,8 @@ def __init__(self, proxy_manager: ProxyManager):
         super().__init__()
         self.proxy_manager = proxy_manager
 
-    def _try_get_channel(self, client_id):
+    def _try_get_channel(self,
+                         client_id: str) -> Optional["grpc._channel.Channel"]:
```
Comment (ijrsvt) [inline python/ray/util/client/server/proxier.py:492]: Thanks!!!

**Review by ijrsvt — APPROVED** [review]
This looks good overall--I think we should validate this with some tests (E2E).

**Comment by ijrsvt** [comment]
@ckw017 Could you try to make a unit test with mocking out the KeepAlive on the server side, like: 
https://github.com/ray-project/ray/blob/5571d2436da42438c26967a5d1211a2ebbeffe42/python/ray/tests/test_client_terminate.py#L93

**Comment by bveeramani** [comment]
## :bangbang: ACTION REQUIRED :bangbang:

We've switched our code formatter from YAPF to Black (see #21311).

To prevent issues with merging your code, here's what you'll need to do:

1. Install Black
```
pip install -I black==21.12b0
```

2. Format changed files with Black
```
curl -o format-changed.sh https://gist.githubusercontent.com/bveeramani/42ef0e9e387b755a8a735b084af976f2/raw/7631276790765d555c423b8db2b679fd957b984a/format-changed.sh
chmod +x ./format-changed.sh
./format-changed.sh
rm format-changed.sh
```

3. Commit your changes.
```
git add --all
git commit -m "Format Python code with Black"
```

4. Merge master into your branch.
```
git pull upstream master
```

5. Resolve merge conflicts (if necessary).

After running these steps, you'll have the updated `format.sh`.

**Comment by ckw017** [comment]
Closing, since manual keep alive pings were ultimately not needed.

**Pull request closed without merging**